### PR TITLE
RavenDB-16926/RavenDB-17846 - RemoveEntryFromRaftLogTest and RemoveEntryFromRaftLogEP fail.

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-16926/SlowTests.Issues.RavenDB15900.RemoveEntryFromRaftLogTest
https://issues.hibernatingrhinos.com/issue/RavenDB-17846/SlowTests.Issues.RavenDB15900.RemoveEntryFromRaftLogEP

### Additional description

the tests `RemoveEntryFromRaftLogTest` and `RemoveEntryFromRaftLogEP` fail.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

**Comment by karmeli87** [comment]
Please apply those changes on a rebased branch to avoid 
image

File: test/SlowTests/Issues/RavenDB-15900.cs [hunk 1]
```diff
@@ -167,83 +169,102 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
...
-                });
+                Assert.True(res);
+            }
 
-                foreach (var server in Servers)
+            
+            foreach (var server in Servers)
+            {
+                long lastCommittedIndex = 0;
+                var res = await WaitForValueAsync(async () =>
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15900.cs]: fix the warning

File: test/SlowTests/Issues/RavenDB-15900.cs [hunk 2]
```diff
@@ -167,83 +169,102 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
...
-                            server.ServerStore.Engine.GetLastCommitIndex(context, out index2, out long term);
-
-                        }
-
-                        return index2 > index;
-                    }, true);
+                        var lastIndex = Cluster.LastRaftIndexForCommand(nodes[0], commandType);
+                        index = lastIndex;
+                    }
+                    catch (TrueException)
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15900.cs]: what is that?

File: test/SlowTests/Issues/RavenDB-15900.cs [hunk 3]
```diff
@@ -167,83 +169,102 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
...
+            }
 
-                    Assert.True(res);
-                }
+        }
 
-                long index2 = 0;
-                foreach (var server in Servers)
+        private async Task<long> WaitForCommandIndex(List<RavenServer> nodes, string commandType, long timeout = 15_000)
+        {
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15900.cs]: there is nothing `async` here

File: test/SlowTests/Issues/RavenDB-15900.cs [hunk 4]
```diff
@@ -167,83 +169,102 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
...
+            await ActionWithLeader(l => l.ServerStore.Engine.StateMachine.Validator = new TestCommandValidator());
 
-                var documentDatabase = await leader.ServerStore.DatabasesLandlord.TryGetOrCreateResourceStore(database);
+            // Stuck leader on this command
+            var testCmd = new RachisConsensusTestBase.TestCommandWithRaftId("test", RaftIdGenerator.NewId());
+            _ = leader.ServerStore.Engine.CurrentLeader.PutAsync(testCmd, TimeSpan.FromSeconds(2));
 
-                var cmd = new RachisConsensusTestBase.TestCommandWithRaftId("test", RaftIdGenerator.NewId());
+            // Get last raft index from leader
+            var testCmdIndex = await WaitForCommandIndex(nodes, "TestCommandWithRaftId");
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15900.cs]: why not to use the existing `LastRaftIndexForCommand` ?

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: test/SlowTests/Issues/RavenDB-15900.cs (hunk 4 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15900.cs]: I use it inside `WaitForCommandIndex`

File: test/SlowTests/Issues/RavenDB-15900.cs (hunk 2 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15900.cs]: `LastRaftIndexForCommand` use `Assert.True` method, and throws `TrueException` if doesn't find the command in the raft log.

File: test/SlowTests/Issues/RavenDB-15900.cs (hunk 1 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15900.cs]: Done

File: test/SlowTests/Issues/RavenDB-15900.cs (hunk 3 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15900.cs]: I forgot to use `Task.Delay` in the loop, that's why I made it async in the first place..
I added `Task.Delay` to the loop.

File: test/SlowTests/Issues/RavenDB-15900.cs [hunk 5]
```diff
@@ -167,90 +167,103 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
...
+                Assert.True(val);
+            }
 
-                    Assert.True(res);
-                }
+        }
 
-                long index2 = 0;
-                foreach (var server in Servers)
+        private async Task<long> WaitForCommandIndex(List<RavenServer> nodes, string commandType, long timeout = 15_000, long interval = 100)
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15900.cs]: why not to use `WaitForRaftCommandToBeAppliedInClusterAsync` ? it does the same thing, no?

Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15900.cs]: Because the command is never applied..

Comment (karmeli87) [reply test/SlowTests/Issues/RavenDB-15900.cs]: so let's create a proper method out of this 
call it `WaitForRaftCommandToBeAppendedInClusterAsync` and move it to be under `ClusterTestBase2`

File: test/SlowTests/Issues/RavenDB-15900.cs (hunk 2 above)
Comment (karmeli87) [reply test/SlowTests/Issues/RavenDB-15900.cs]: I don't like the `TrueException`, let's catch here any exception or maybe even better let's have a method called 
`WaitForRaftCommandToBeAppendedAsync` based on the `LastRaftIndexForCommand`

File: test/SlowTests/Issues/RavenDB-15900.cs (hunk 5 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15900.cs]: Done

File: test/Tests.Infrastructure/RavenTestBase.Cluster.cs [hunk 6]
```diff
@@ -45,6 +46,37 @@ public async Task WaitForRaftCommandToBeAppliedInLocalServerAsync(string command
...
+            while (sw.ElapsedMilliseconds < timeout)
+            {
+                foreach (var server in nodes)
+                {
+                    try
+                    {
+                        var lastIndex = LastRaftIndexForCommand(server, commandType);
+                        index = lastIndex;
+                    }
+                    catch (TrueException)
```
Comment (karmeli87) [inline test/Tests.Infrastructure/RavenTestBase.Cluster.cs]: `TrueException` is a xunit exception, we don't usually catch those.
Also what if we waited for the 15sec we want to propogate the last exception (no matter what it was)
similar to `WaitForValueAsync`.
Also let's throw if we didn't find this on time

Comment (shaharhikri) [reply test/Tests.Infrastructure/RavenTestBase.Cluster.cs]: Done

**Review by karmeli87 — APPROVED** [review]

**Pull request merged**